bee-ledger/src/workers/consensus/white_flag.rs

```diff
@@ -68,54 +58,190 @@ fn apply_regular_essence<B: StorageBackend>(
                     return Ok(ConflictReason::InputUtxoNotFound);
                 }
             }
-            Input::Treasury(_) => {
+            _ => {
                 return Err(Error::UnsupportedInputKind(input.kind()));
             }
         };
 
-        let _essence_hash = TransactionEssence::from(essence.clone()).hash();
+        consumed_outputs.push((*output_id, consumed_output));
+    }
 
-        match consumed_output.inner() {
-            Output::Treasury(_) => return Err(Error::UnsupportedOutputKind(consumed_output.inner().kind())),
-            Output::Basic(_) => {
-                // TODO
-            }
-            Output::Alias(_) => {
-                // TODO
-            }
-            Output::Foundry(_) => {
-                // TODO
+    let inputs: Vec<(OutputId, &Output)> = consumed_outputs
+        .iter()
+        .map(|(output_id, created_output)| (*output_id, created_output.inner()))
+        .collect();
+
+    let mut context = ValidationContext::new(
+        transaction_id,
+        essence,
+        inputs.iter().map(|(output_id, input)| (output_id, *input)),
```

Comment:
that's essentially copying the inputs? Why not create the `inputs` Vec with owned `OutputId` and `Output`. You deref the `input` here anyways.

Reply:
`Output` is not `Copy` so no it's not copying the inputs. We are creating an iterator over a Vec of `&Output` so the type ends up being `&&Output`, just removing one ref here.